### PR TITLE
Reworked internals

### PR DESCRIPTION
**Pull request by RobinGeens**
This branch holds many many changes to the internals of Zigzag, though the functionality itself has barely changed. Changes include the following:

- Fully rewrite (User) Spatial Mapping Generator 
- Force typehints everywhere and test with dynamic type checking
- Represent Layer Dimensions, Operands, Operational Array Dimensions, etc... with classes instead of strings
- Encapsulate most of the Layer Attributes in classes
- Equip a LayerAttributes class with methods to parse user input
- Provide an abstract base class for Workload and CME
- Transfer the functionality of adding CMEs to a `CumulativeCME` class
- Simplify the file structure.

The proposed dev setup, to be used by all future contributions, is as follows:

- Use the .pylintrc file from `.github/workflows/`. This is the one from the Google style guide, but with a line length of 120
- Use strict static type checking: ` "python.analysis.typeCheckingMode": "strict",` in `.vscode/settings.json`
- Test with Typeguard (dynamic type checking).

To complete this PR, the following to do's should be completed. Currently, only the `test_origin` tests pass.

- [ ] Review the new code
- [ ] Repair the following files and provide type annotations:  `PEArrayScalingStage.py`, `SearchUnusedMemoryStage.py`, `RemoveUnusedMemoryStage.py`, `GeneralParameterIteratorStage.py`, `AimcArray.py`, `DimcArray.py`, `imc_unit.py`, `ImcArray.py`
- [ ] Run the tests, other than `test_origin`
- [ ] Replace the hardcoded energy and latency values in those tests. With the new SpatialMappingGenerator, the results are slightly better
- [ ] Test whether CACTI runs as expected
- [ ] Update the docs

To come to a more uniformly programmed code base, I would like to suggest the following programming principles:

1. A good function should do one thing and one thing only. As such, a good function cannot reasonably be separated into multiple blocks, as this implies the function does more than one thing.
2. Function and variable names should be clear enough such that a comment to describe it is redundant. With well-chosen function and variable names, it is easy to follow the programming flow without reading additional comments.
3. Comments should give a rationale, and not explain the code itself.
4. Avoid try/except blocks for control flow, these should only be used for error handling and irregular behavior.

Note that after this PR, I will be reworking the user interface.

### CODE REVIEW

File: .github/workflows/python-test.yml [hunk 1]
```diff
@@ -1,12 +1,7 @@
 name: Python test
 
 # Controls when the workflow will run
-on:
-  # Triggers the workflow on push or pull request events but only for the "master" branch
-  push:
-    branches: [ "master" ]
-  pull_request:
-    branches: [ "master" ]
+on: [push]
```
Comment (jorendumoulin) [inline .github/workflows/python-test.yml:4]: why this change? Checks on pull requests are very nice

File: .github/workflows/python-test.yml [hunk 2]
```diff
@@ -43,4 +38,4 @@ jobs:
       - name: Test with pytest
         run: |
           pip install pytest
-          pytest tests/
\ No newline at end of file
+          python -m pytest --typeguard-packages=zigzag tests/
```
Comment (jorendumoulin) [inline .github/workflows/python-test.yml]: please add a newline to every file you save :) 
https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files

File: tests/main/test_origin/test_ascend_like.py [hunk 3]
```diff
@@ -11,10 +11,10 @@
...
 ens_lats = {
-    "zigzag/inputs/examples/workload/alexnet.onnx": (5738192980.375, 8728331),
-    "zigzag/inputs/examples/workload/mobilenetv2.onnx": (1913797698.5250015, 7426499),
-    "zigzag/inputs/examples/workload/resnet18.onnx": (1858697886.165, 3720129),
-    "zigzag.inputs.examples.workload.resnet18": (2408671233.7250004, 4804196),
+    "zigzag/inputs/examples/workload/alexnet.onnx": (5737868753.12, 8696023),
+    "zigzag/inputs/examples/workload/mobilenetv2.onnx": (1913561726.0000005, 7359650),
+    "zigzag/inputs/examples/workload/resnet18.onnx": (1860918012.2400002, 3698589),
+    "zigzag.inputs.examples.workload.resnet18": (2345967030.96, 4779555),
 }
```
Comment (jorendumoulin) [inline tests/main/test_origin/test_ascend_like.py:19]: why do all these numbers change?

File: zigzag/cost_model/port_activity.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+from zigzag.datatypes import LayerOperand
+from zigzag.hardware.architecture.memory_port import DataDirection
+
+
+class PortActivity:
+    """!  Class that collects all the data transfer rate (periodic) information for each DTL (data transfer link)."""
```
Comment (jorendumoulin) [inline zigzag/cost_model/port_activity.py:6]: What are all these exclamation marks doing in the comments?

**Review by jorendumoulin — COMMENTED** [review]
I didn't look at this in detail (and probably wont for 14k changes), but here are some thoughts:

> This branch holds many many changes to the internals of Zigzag, though the functionality itself has barely changed. Changes include the following:
> 
>     Fully rewrite (User) Spatial Mapping Generator
>     Force typehints everywhere and test with dynamic type checking
>     Represent Layer Dimensions, Operands, Operational Array Dimensions, etc... with classes instead of strings
>     Encapsulate most of the Layer Attributes in classes
>     Equip a LayerAttributes class with methods to parse user input
>     Provide an abstract base class for Workload and CME
>     Transfer the functionality of adding CMEs to a CumulativeCME class
>     Simplify the file structure.

This is a perfect list of how this PR had to be split up 🥲 . Right now, if someone disagrees with a certain part, it's very hard to filter it out. It is also very difficult to see what diff corresponds to what feature change.

> The proposed dev setup, to be used by all future contributions, is as follows:
> 
>     Use the .pylintrc file from .github/workflows/. This is the one from the Google style guide, but with a line length of 120
>     Use strict static type checking:  "python.analysis.typeCheckingMode": "strict", in .vscode/settings.json
>     Test with Typeguard (dynamic type checking).

Not everybody uses VSCode :), if you want to enforce type checking, I suggest adding the pyright configuration (which can be detected by VSCode) in the `pyproject.toml`, see https://github.com/xdslproject/xdsl/blob/main/pyproject.toml for an example`. In the configuration, you can set specific paths to exclude/include from type checking to allow for errors in files where it is not needed / types have not been added yet.

For linting and formatting, all the new kids on the block run `ruff`, which is much faster, but I would not know if `pylint` is too slow right now.

If you want help setting up this stuff let me know, I have done this a couple of times by now.

Other than that, all the new class based and typed implementations look very nice 😍

**Comment by RobinGeens** [comment]
> This is a perfect list of how this PR had to be split up 🥲 . Right now, if someone disagrees with a certain part, it's very hard to filter it out. It is also very difficult to see what diff corresponds to what feature change.

Agreed, will keep this in mind next PR :)


> Not everybody uses VSCode :), if you want to enforce type checking, I suggest adding the pyright configuration (which can be detected by VSCode) in the `pyproject.toml`

Fair point, thanks for the remark!

File: .github/workflows/python-test.yml (hunk 1 above)
Comment (RobinGeens) [reply .github/workflows/python-test.yml:4]: I wanted to check on every push, at least for the duration that we are working on this PR. Why don't we test on every push anyways?

File: tests/main/test_origin/test_ascend_like.py (hunk 3 above)
Comment (RobinGeens) [reply tests/main/test_origin/test_ascend_like.py:19]: In the old configuration, when a mapping is provided for a certain operator (e.g. C : 32), but the loop dimension is not present in the layer (so no C loop dimension), the unrolling would be reduced to 1.
In the new configuration, when such an invalid mapping is provided, it gets deleted (with a warning to the user), and replaced with a generated one. Thus, all tests results are slightly better. You can also visually see this: there are no spatial stalls with the new version.

File: zigzag/cost_model/port_activity.py (hunk 4 above)
Comment (RobinGeens) [reply zigzag/cost_model/port_activity.py:6]: Apparently, Doxygen needs them to parse stuff like `@param`. Not sure if this can be resolved in another way.

File: .github/workflows/python-test.yml (hunk 1 above)
Comment (jorendumoulin) [reply .github/workflows/python-test.yml:4]: Ah i misunterstood this I thought it was replaced by only push to main, but it is every push. Ok!

File: tests/main/test_origin/test_ascend_like.py (hunk 3 above)
Comment (jorendumoulin) [reply tests/main/test_origin/test_ascend_like.py:19]: Aha, interesting

File: zigzag/cost_model/port_activity.py (hunk 4 above)
Comment (jorendumoulin) [reply zigzag/cost_model/port_activity.py:6]: https://www.doxygen.nl/manual/docblocks.html#pythonblocks
The other way is setting `PYTHON_DOCSTRING NO` in the configuration.
This is fine for me too though, I just had never seen it before and it doesn't seem very common

Comment (asyms) [reply zigzag/cost_model/port_activity.py:6]: So the docstring can be without the exclamation mark? 

If so, @RobinGeens could you please update the PR as such?

File: tests/main/test_origin/test_ascend_like.py (hunk 3 above)
Comment (asyms) [reply tests/main/test_origin/test_ascend_like.py:19]: @RobinGeens can you update the numbers for all working tests?

For the sake of progress, we can remove the other tests of files (they will be updated by @JiacongSun later).

Comment (RobinGeens) [reply tests/main/test_origin/test_ascend_like.py:19]: The numbers have been updated for all `test_origin ` tests. The others rely on Jiacong's code

File: zigzag/cost_model/port_activity.py (hunk 4 above)
Comment (RobinGeens) [reply zigzag/cost_model/port_activity.py:6]: Hmm I can't seem to get this working at all. For most classes/functions, the docstring does not show up in the html, regardless of the exclamation mark or the `PYTHON_DOCSTRING NO` setting. Anyone else want to have a try at this?

**Review by asyms — APPROVED** [review]
LGTM (apart from imc architecture and lower memory level removal stages which will be updated by @JiacongSun in a future PR).
